src/s3ql/backends/backblaze.py

```diff
@@ -0,0 +1,704 @@
+            raise ValueError('max_filecount maximum is 1000')
+
+        request_dict = dict(bucketId=self.bucket_id, maxFileCount=max_filecount)
+        request_dict['startFileName'] = self.prefix + self._encode_key(key)
+
+        # supposing there is less than 1000 old file version
+        body = bytes(json.dumps(request_dict), encoding='UTF-8')
+
+        resp, body = self._do_request('POST',
+                                      '/b2api/v1/b2_list_file_versions',
+                                      self.conn_api,
+                                      body=body)
+
+        j = json.loads(str(body, encoding='UTF-8'))
+        r = []
+
+        # We suppose there is less than 1000 file version
+        for f in j['files']:
+            if self._decode_key(f['fileName']) == self.prefix + key:
+                r.append({'fileName': f['fileName'],
+                          'fileId': f['fileId']})
+        return r
+
+    @retry
+    def _list_file_name(self, start_filename=None, max_filecount=1000):
+        if max_filecount > 1000:
+            raise ValueError('max_filecount maximum is 1000')
+
+        request_dict = {'bucketId': self.bucket_id,
+                        'maxFileCount': max_filecount}
+        if start_filename is not None:
+            request_dict['startFileName'] = start_filename
+        body = bytes(json.dumps(request_dict), encoding='UTF-8')
+
+        resp, body = self._do_request('POST',
+                                      '/b2api/v1/b2_list_file_names',
+                                      self.conn_api,
+                                      body=body)
+
+        j = json.loads(str(body, encoding='UTF-8'))
+        filelist = [f['fileName'] for f in j['files']]
+
+        return j['nextFileName'], filelist
+
+    def _encode_key(self, filename):
+
+        # URLencode filename
+        filename = urllib.parse.quote(filename.encode('utf-8'), safe='/\\')
+
+        # escape backslashes as backend does not support them
+        filename = filename.replace("\\","=5C")
+
+        return filename
+
+    def _decode_key(self,filename):
+
+        # unescape backslashes as backend does not support them
+        filename = filename.replace("=5C","\\")
+
+        # URLencode filename
+        filename = urllib.parse.unquote(filename)
+
+        return filename
+
+    def has_native_rename(self):
+        """True if the backend has a native, atomic rename operation"""
+        return False
+
+    @copy_ancestor_docstring
+    def is_temp_failure(self, exc):
+
+        if isinstance(exc, AuthenticationExpired):
+            return True
+
+        if isinstance(exc, ssl.SSLError):
+            return True
+
+        if is_temp_network_error(exc):
+            return True
+
+        if isinstance(exc, HTTPError):
+            return True
+
+        if isinstance(exc, B2Error):
+            if (exc.status == 400 or \
+                exc.status == 408 or \
+                (exc.status >= 500 and exc.status <= 599)):
```

Comment:
Especially 4xx error codes are reserved for "client errors" - rather unlikely to be temporary. Do you have a pointer to documentation that says these should be retried?